### PR TITLE
Add support for OLMo's November release

### LINKED ISSUE #34496: Add support for OLMo November release
### Model description

An updated OLMo model will be released in November. The new model has a few small architecture changes compared to the existing model in transformers:

- RMSNorm is used instead of standard layer norm.
- Norm is applied to attention queries and keys.
- Norm is applied after attention/feedforward rather than before.

It would be great to have it integrated in transformers from day 1.

### Open source status

- [X] The model implementation is available
- [ ] The model weights are available

### Provide useful links for the implementation

OLMo code repo: https://github.com/allenai/OLMo
Authors include @2015aroras 
Weights are not yet available, will be made available at release.

### PR DESCRIPTION
**Pull request by 2015aroras**
# What does this PR do?

An updated OLMo model will be released in November. The new model has a few small architecture changes compared to the existing model in transformers:

- RMSNorm is used instead of standard layer norm.
- Norm is applied to attention queries and keys.
- Norm is applied after attention/feedforward rather than before.

This PR updates the OLMo implementation in transformers to support the November release.

@ArthurZucker 

Fixes #34496 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by 2015aroras** [comment]
Tested that an intermediate checkpoint can be converted and matches the original model (max logit diff <5e-5).

Tested that a random [OLMo 1B](https://huggingface.co/allenai/OLMo-1B-hf) checkpoint can still be converted and still matches the original model (max logit diff <5e-5).

File: src/transformers/models/olmo/modeling_olmo.py [hunk 1]
```diff
@@ -238,6 +270,16 @@ def __init__(self, config: OlmoConfig, layer_idx: Optional[int] = None):
...
+        self.q_norm = (
+            get_layer_norm(config.layer_norm_type, self.num_heads * self.head_dim, config.rms_norm_eps)
+            if config.use_q_norm
+            else None
+        )
+        self.k_norm = (
+            get_layer_norm(config.layer_norm_type, self.num_key_value_heads * self.head_dim, config.rms_norm_eps)
+            if config.use_k_norm
+            else None
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/olmo/modeling_olmo.py:282]: hey! This goes a bit against the `transformers` philosophy: we never change an old model to support a new architecture! We need a `Olmo2` model 🤗 
with modular this should be fairly simple to implement!

**Review by ArthurZucker — COMMENTED** [review]
Thanks for the pr! 
We should create a new model with https://huggingface.co/docs/transformers/en/modular_transformers 🤗

File: src/transformers/models/olmo/modeling_olmo.py (hunk 1 above)
Comment (2015aroras) [reply src/transformers/models/olmo/modeling_olmo.py:282]: Ok, I've re-implemented the model as a new model using modular. I'll put out the PR once we have decided on a suitable model name internally (we don't intend to call this release OLMo 2).

**Comment by 2015aroras** [comment]
Making a fresh PR with modular transformers.

**Pull request closed without merging**